### PR TITLE
[WIP] Added autofill functionality

### PR DESCRIPTION
**Pull request by ganttastic**
## Description

This PR adds autofill functionality to text fields by adding `TextContentType`. This informs the platform OS what should be filled in.

## Related Issues

This is related to: Issue #13015 

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*
- [ ] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/services/text_input.dart [hunk 1]
```diff
@@ -407,6 +407,72 @@ enum TextCapitalization {
   none,
 }
 
+/// Configures how the platform will offer auto-fill suggestions for the input
+enum TextContentType {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart]: Both [UITextContentType](https://developer.apple.com/documentation/uikit/uitextcontenttype) and Android autofill hints are raw strings under the hood, and you can create them using arbitrary strings. `enum` is a closed type. Should we just use `String` instead or use some wrapper class that allows for custom content types?

**Review by LongCatIsLooong — COMMENTED** [review]
Hi @ganttastic , Thank you for the pull request! 

I'm currently working on autofill as well. On iOS if you put a username text field and a password text field in the same view controller they will both get autofilled in one go, and this is even more common on Android and web. I'm currently drafting a design doc for implementing such feature, and hopefully it will be published soon. In the meantime as always ideas or suggestions are welcome!

File: packages/flutter/lib/src/services/text_input.dart (hunk 1 above)
Comment (ganttastic) [reply packages/flutter/lib/src/services/text_input.dart]: 👍 I like the idea of using something closer to the raw value. That would make it easier to maintain as they continue to add content types.

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -407,6 +407,72 @@ enum TextCapitalization {
...
+  /// Accepts full name
```
[the remark this answers is not in the document]
Comment (ganttastic) [reply packages/flutter/lib/src/services/text_input.dart]: 👍 Will add those to the comments.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -522,6 +523,11 @@ class EditableText extends StatefulWidget {
   /// {@endtemplate}
   final bool autocorrect;
 
+  /// Specifies the expected input type to inform the platform's auto-fill system
```
[the remark this answers is not in the document]
Comment (ganttastic) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Nice! I’ll update that.

**Comment by ganttastic** [comment]
@LongCatIsLooong Thank you for the feedback!

I don’t want to step on your toes here if you already have stuff in place. Is my PR conflicting with what you’re working on? Happy to pivot and help you in whatever way I can!

File: packages/flutter/lib/src/services/text_input.dart (hunk 1 above)
Comment (ganttastic) [reply packages/flutter/lib/src/services/text_input.dart]: @LongCatIsLooong I've updated that enum to a custom class. Let me know if this is what you had in mind.

File: packages/flutter/lib/src/services/text_input.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (ganttastic) [reply packages/flutter/lib/src/services/text_input.dart]: I've added the iOS and Android definitions.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (ganttastic) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Updated to use comment template macros.

**Comment by LongCatIsLooong** [comment]
Sorry for the hold up! I'm finishing up a design proposal and a proof-of-concept implementation, the ETA is either this week or next. Would definitely like to get your input if you're still interested!

**Comment by ganttastic** [comment]
@LongCatIsLooong No worries! I'm still interested and would love to help in whatever way I can.

**Comment by ganttastic** [comment]
Much more robust solution provided in PR #52126. Looks great!

**Pull request closed without merging**